### PR TITLE
OMG-73 Optimize the merkle root calculation for fixed depth Merkles

### PR DESCRIPTION
**Pull request by pgebal**

### CODE REVIEW

**Comment by paulperegud** [comment]
```
lib/merkle_tree/proof.ex:86: Invalid type specification for function 'Elixir.MerkleTree.Proof':path_from_binary/1. The success typing is (<<_:1,_:_*1>>) -> {0 | 1,bitstring()}
```

File: lib/merkle_tree.ex [hunk 1]
```diff
@@ -42,35 +45,55 @@ defmodule MerkleTree do
...
+
+  defp pow(n), do: :math.pow(2, n) |> round
+
   @doc """
     Builds a new binary merkle tree.
   """
-  @spec build(blocks, hash_function) :: root
-  def build(blocks, hash_function) do
-    starting_height = 0
+  def build(blocks, hash_function, max_height) when blocks != [] do
```
Comment (paulperegud) [inline lib/merkle_tree.ex]: when doing recursive calls, place base case (`def build([],...)`) at the top

File: lib/merkle_tree.ex [hunk 2]
```diff
@@ -22,10 +22,13 @@ defmodule MerkleTree do
...
   defstruct [:blocks, :root, :hash_function]
 
-  @number_of_children 2 # Number of children per node. Configurable.
+  @number_of_children 2 # Number of children per node
+  @default_height 16
+  @zeroes <<0>> |> List.duplicate(32) |> Enum.join
 
   @type blocks :: [String.t, ...]
   @type hash_function :: (String.t -> String.t)
+  @type height :: non_neg_integer
```
Comment (paulperegud) [inline lib/merkle_tree.ex:30]: Height is number, not ordinal; so always positive and minimal value is 1

Comment (paulperegud) [reply lib/merkle_tree.ex:30]: I.e. MerkleTree.new(..,.., 0) should crash

File: lib/merkle_tree.ex [hunk 3]
```diff
@@ -22,10 +22,13 @@ defmodule MerkleTree do
 
   defstruct [:blocks, :root, :hash_function]
 
-  @number_of_children 2 # Number of children per node. Configurable.
+  @number_of_children 2 # Number of children per node
```
Comment (paulperegud) [inline lib/merkle_tree.ex]: Since this functionality does not work at all anyway, maybe just drop it?

Comment (paulperegud) [reply lib/merkle_tree.ex]: ~~We've already broken compatibility with original implementation by introducing fixed height.~~

EDIT: scratch that

File: lib/merkle_tree.ex [hunk 4]
```diff
@@ -42,35 +45,55 @@ defmodule MerkleTree do
...
     ``(String.t -> String.t)``.
   """
-  @spec new(blocks, hash_function) :: t
-  def new(blocks, hash_function \\ &MerkleTree.Crypto.sha256/1)
-  when blocks != [] do
-    unless is_power_of_n(@number_of_children, Enum.count(blocks)), do: raise MerkleTree.ArgumentError
-
-    root = build(blocks, hash_function)
+  @spec new(blocks, hash_function, height) :: t
+  def new(blocks, hash_function \\ &MerkleTree.Crypto.sha256/1, height \\ @default_height) do
```
Comment (paulperegud) [inline lib/merkle_tree.ex]: if length(blocks) > 2^height, then it enters never-ending loop

**Review by paulperegud — CHANGES_REQUESTED** [review]
Please also update README.md

File: lib/merkle_tree.ex (hunk 2 above)
Comment (pgebal) [reply lib/merkle_tree.ex:30]: As we speaked, let's leave it this way.

File: lib/merkle_tree.ex (hunk 1 above)
Comment (pgebal) [reply lib/merkle_tree.ex]: done

File: lib/merkle_tree.ex (hunk 2 above)
Comment (paulperegud) [reply lib/merkle_tree.ex:30]: (height in merkle tree is defined as number of edges on path between leaf and root; my bad)

File: lib/merkle_tree.ex (hunk 3 above)
Comment (pgebal) [reply lib/merkle_tree.ex]: done

File: lib/merkle_tree.ex (hunk 4 above)
Comment (pgebal) [reply lib/merkle_tree.ex]: done

**Review by paulperegud — APPROVED** [review]
LGTM

**Pull request merged**